### PR TITLE
[WIP] Add TypeScript documentation

### PR DESCRIPTION
**Pull request by seamc**
TypeScript + antd via babel.  Issue #5246

### CODE REVIEW

**Comment by benjycui** [comment]
ref: https://github.com/ant-design/ant-design/issues/5426 ?

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+[create-react-app](https://github.com/facebookincubator/create-react-app) is one of best React application development tool, we are going to use `antd` within it and modify the webpack config for some customized needs.
+
+---
+
+## Install and Initialization
+
+We need to install `create-react-app` first, you may need install [yarn](https://github.com/yarnpkg/yarn/) too.
+
+Additionally, install [react-scripts-ts](https://github.com/wmonk/create-react-app-typescript)
```
Comment (bang88) [inline docs/react/use-with-create-react-app-ts.en-US.md]: No need install this, right ?
all you need is specify the scripts version.
```sh
--scripts-version=react-scripts-ts 
```

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+```bash
+$ npm install -g create-react-app yarn react-scripts-ts
```
Comment (bang88) [inline docs/react/use-with-create-react-app-ts.en-US.md]: same here.

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+│──package.json
+│──README.md
+│──tsconfig.json
+│──tslint.json
+│──yarn.lock
+├── public
+│──├── favicon.ico
+│──└── index.html
+├──src
+│──└──App.css
```
Comment (infeng) [inline docs/react/use-with-create-react-app-ts.en-US.md]: Lack of space.

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+   },
+
+...
+
++ // Parse less files and modify variables
++ {
++   test: /\.less$/,
++   loader: 'style!css!postcss!less?{modifyVars:{"@primary-color":"#1DA57A"}}'
++ },
+]
```
Comment (afc163) [inline docs/react/use-with-create-react-app-ts.en-US.md]: https://github.com/ant-design/ant-design/commit/be238c788775990d4f15c9f83c16ca4db7226271

Need to update there for cra 1.x

**Comment by benjycui** [comment]
ping @seamc

**Comment by seamc** [comment]
Thanks for the feedback!  Looks good with the commits provided.

**Comment by colinshen** [comment]
warning after using latest react v1.5.*, I'm not sure this warning caused by ant because ,in a new project created by[ create-react-app ts](https://github.com/wmonk/create-react-app-typescript), there are no warnnings. you guys may look at the [configs  ](https://github.com/ant-design/ant-design/pull/6193/files) and update it to have a better experiences of using ant design with typescript. However, I tried to follow this documentation, but we still need to add babel-loader for importing components on demand, which doesn't shows in the documentation.
```
Warning: Accessing PropTypes via the main React package is deprecated, and will be removed in  React v16.0. Use the latest available v15.* prop-types package from npm instead. For info on usage, compatibility, migration and more, see https://fb.me/prop-types-docs
Warning: Accessing createClass via the main React package is deprecated, and will be removed in React v16.0. Use a plain JavaScript class instead. If you're not yet ready to migrate, create-react-class v15.* is available on npm as a temporary, drop-in replacement. For more info see https://fb.me/react-create-class
```
My configs
```
//tsconfig.js
{
  "compilerOptions": {
    "outDir": "build/dist",
    "module": "es6", // use ES2015 modules
    "target": "es6", // compile to ES2015 (Babel will do the rest)
    "lib": ["es6", "dom"],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "preserve",
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "rootDir": "src",
    "forceConsistentCasingInFileNames": false,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": true,
    "removeComments": true
  },
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "src/setupTests.ts"
  ],
  "types": [
    "typePatches"
  ]
}
```
**add babel-loader**. In my project, without this loader, the project has compile error, which may caused by preserve setting in ts config.
```
//webpack.dev.config
//only tsx config
// Compile .tsx?
      {
        test: /\.(ts|tsx)$/,
        include: paths.appSrc,
        loader:  [require.resolve('babel-loader'), require.resolve('ts-loader')]
      }
```
```
//.babelrc
{
    "presets": ["react", "es2015"],
    "plugins": [["import", {"libraryName": "antd", "style": true }]] 
}
```

**Comment by benjycui** [comment]
ping~ @seamc

**Comment by seamc** [comment]
create-react-app and wmonk's typescript scripts have been updated and I will submit a new pull request to update the documentation.

Concerning the warnings: the issues in #5678 may have something to do with it or it might be babel.  Typescript compilation alone does not bring up deprecation warnings; they only pop up when antd components are imported via babel (babel/babel#5705 ? maybe).

File: docs/react/use-with-create-react-app-ts.en-US.md (hunk 4 above)
Comment (colinshen) [inline docs/react/use-with-create-react-app-ts.en-US.md]:
```
{
        test: /\.(ts|tsx)$/,
        include: paths.appSrc,
        use:[
          require.resolve('babel-loader'),
          require.resolve('ts-loader')
        ]
      },
```

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+### Import on demand
+
+[babel-plugin-import](https://github.com/ant-design/babel-plugin-import) is a babel plugin for importing components on demand ([principle](/docs/react/getting-started#Import-on-Demand)). After ejecting all config files to antd-demo, we are allowed to modify `config/webpack.config.dev.js`.
+
+Note that the following setup will effectively compile your TypeScript with ES6 module resolution and then ship it over to Babel-Loader to be transpiled.
+
+```bash
+$ yarn add babel-core babel-loader babel-plugin-import babel-plugin-import babel-preset-react babel-preset-es2015 --save-dev
+```
```
Comment (comerc) [inline docs/react/use-with-create-react-app-ts.en-US.md]: typo: babel-plugin-import babel-plugin-import

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+Update the webpack loader configuration for TypeScript files.  `config/webpack.config.dev.js`
+
+```diff
+{
+  test: /\.(ts|tsx)$/,
+  include: paths.appSrc,
++  loader: 'babel!ts'
+-  loader: 'ts',
+},
```
Comment (comerc) [inline docs/react/use-with-create-react-app-ts.en-US.md]: ??? how to replace:
```
loader: require.resolve('ts-loader'),
```

File: docs/react/use-with-create-react-app-ts.en-US.md (hunk 1 above)
Comment (abenhamdine) [reply docs/react/use-with-create-react-app-ts.en-US.md]: you're right

File: docs/react/use-with-create-react-app-ts.en-US.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+.App {
+  text-align: center;
+}
+
+...
+```
+
+Ok, you now see a blue primary button displaying in page now, next you can choose any components of `antd` to develop your application. Visit other workflow of `create-react-app` at its [User Guide ](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md).
+
```
Comment (abenhamdine) [inline docs/react/use-with-create-react-app-ts.en-US.md]: All previous steps work perfectly, but at this step, I run into this error : 
```
C:\node-projects\antd-demo-ts\node_modules\antd\lib\upload\interface.d.ts
(2,8): error TS1192: Module '"C:/node-projects/antd-demo-ts/node_modules/@types/react/index"' has no default export.
```

**Review by abenhamdine — CHANGES_REQUESTED** [review]

**Comment by abenhamdine** [comment]
IMHO, this whole PR should be revisited now it's possible to use cra without eject.
I successfully made it work by following both https://github.com/comerc/cra-ts-antd and https://ant.design/docs/react/use-with-create-react-app#Advanced-Guides.
You can see the result here : https://github.com/abenhamdine/ant-demo-ts-cra

File: docs/react/use-with-create-react-app-ts.en-US.md (hunk 7 above)
Comment (apieceofbart) [reply docs/react/use-with-create-react-app-ts.en-US.md]: I've got a different error (after running up to this point):
```
/home/user/projects/antd-demo-ts/node_modules/antd/lib/upload/interface.d.ts
(2,8): error TS1192: Module '"/home/user/projects/antd-demo-ts/node_modules/@types/react/index"' has no default export.
```
This is fixed by adding `"allowSyntheticDefaultImports": true` to tsconfig

Comment (abenhamdine) [reply docs/react/use-with-create-react-app-ts.en-US.md]: yes the issue I mentionned is fixed, but `allowSyntheticDefaultImports` is still required, I've added it at https://github.com/abenhamdine/ant-demo-ts-cra/blob/master/README.md#step-7--edit-tsconfigjson

**Comment by afc163** [comment]
@abenhamdine Great job, I think we should close this PR and create new one, could you help us finish the documentation?

**Comment by abenhamdine** [comment]
Thx ! I'd be glad to send a new PR, but I'm currently struggling to make work together react-app-rewired and last version of create-react-app-typescript 2.7.0 (probably because of webpack@3 and awesome typescript loader).
I will let you know if I succeed in fixing this.
If not, I can still create a PR but people may be deceived because they are likely to install last version of cra-ts and to be stuck as well.

**Pull request closed without merging**